sky/backends/cloud_vm_ray_backend.py

```diff
@@ -2714,15 +2741,30 @@ def _exec_code_on_head(
         else:
             job_submit_cmd = (
                 'RAY_DASHBOARD_PORT=$(python -c "from sky.skylet import job_lib; print(job_lib.get_job_submission_port())" 2> /dev/null || echo 8265);'  # pylint: disable=line-too-long
-                f'{cd} && mkdir -p {remote_log_dir} && ray job submit '
+                f'{cd} && ray job submit '
                 '--address=http://127.0.0.1:$RAY_DASHBOARD_PORT '
                 f'--submission-id {ray_job_id} --no-wait '
                 f'"{executable} -u {script_path} > {remote_log_path} 2>&1"')
 
+            mkdir_code = (f'{cd} && mkdir -p {remote_log_dir} && '
+                          f'touch {remote_log_path}')
+            code = job_lib.JobLibCodeGen.queue_job(job_id, job_submit_cmd)
+            job_submit_cmd = mkdir_code + ' && ' + code
+
         returncode, stdout, stderr = self.run_on_head(handle,
                                                       job_submit_cmd,
                                                       stream_logs=False,
                                                       require_outputs=True)
+
+        if 'has no attribute' in stdout:
+            # Happens when someone calls `sky exec` but remote is outdated
+            # necessicating calling `sky launch`
+            with ux_utils.print_exception_no_traceback():
+                raise RuntimeError(
+                    f'{colorama.Fore.RED}SkyPilot runtime is stale on the '
+                    'remote cluster. To update, run: sky launch -c '
+                    f'{handle.cluster_name}{colorama.Style.RESET_ALL}')
```

Comment:
Ran into this when testing #1890 smoke tests (commit b020963a176ff7a07c7c090c14502df927f5a799) using a smoke VM launched by `sky`. I added code to print out stdout:
```
RuntimeError: SkyPilot runtime is stale on the remote cluster. To update, run: sky launch -c t-env-check-3917-b8
output:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'sky.skylet.job_lib' has no attribute 'scheduler'
```
This test cluster was freshly launched from the smoke VM, so not sure why the initial `launch` followed by an `exec` would trigger this staleness error?

cc @Michaelvll

Reply:
Btw, is this check necessary for the PR to function? The `'has no attribute' in stdout` can be brittle.